components/tidb_query/src/codec/mysql/time/mod.rs

```diff
@@ -57,605 +48,1156 @@ const MONTH_NAMES_ABBR: &[&str] = &[
+    pub fn micro(self) -> u32 {
+        self.get_micro()
     }
 
-    pub fn is_zero(&self) -> bool {
-        self.time.timestamp() == ZERO_TIMESTAMP
+    /// Returns the year number
+    pub fn year(self) -> u32 {
+        self.get_year()
     }
 
-    pub fn invalid_zero(&self) -> bool {
-        self.time.month() == 0 || self.time.day() == 0
+    /// Returns the month number
+    pub fn month(self) -> u32 {
+        self.get_month()
     }
 
-    pub fn get_fsp(&self) -> u8 {
-        self.fsp
+    /// Returns the day number
+    pub fn day(self) -> u32 {
+        self.get_day()
     }
+}
 
-    pub fn set_fsp(&mut self, fsp: u8) {
-        self.fsp = fsp;
+mod parser {
+    use super::*;
+    fn bytes_to_u32(input: &[u8]) -> Option<u32> {
+        input.iter().try_fold(0u32, |acc, d| {
+            d.is_ascii_digit().as_option()?;
+            acc.checked_mul(10)
+                .and_then(|t| t.checked_add(u32::from(d - b'0')))
+        })
     }
 
-    pub fn get_time(&self) -> DateTime<Tz> {
-        self.time
+    fn digit1(input: &[u8]) -> Option<(&[u8], &[u8])> {
+        let end = input
+            .iter()
+            .position(|&c| !c.is_ascii_digit())
+            .unwrap_or_else(|| input.len());
+        (end != 0).as_option()?;
+        Some((&input[end..], &input[..end]))
     }
 
-    pub fn set_time(&mut self, time: DateTime<Tz>) {
-        self.time = time
+    fn space1(input: &[u8]) -> Option<&[u8]> {
+        let end = input
+            .iter()
+            .position(|&c| !c.is_ascii_whitespace())
+            .unwrap_or_else(|| input.len());
+
+        (end < input.len()).as_option()?;
+        Some(&input[end..])
     }
 
-    /// Converts a `DateTime` to printable string representation
-    #[inline]
-    pub fn to_numeric_string(&self) -> String {
-        if self.time_type == TimeType::Date {
-            if self.is_zero() {
-                String::from(ZERO_DATE_NUMERIC_STR)
-            } else {
-                format!("{}", self.time.format("%Y%m%d"))
-            }
-        } else {
-            if self.is_zero() {
-                if self.fsp > 0 {
-                    // Do we need to round the result?
-                    let nanos = self.time.nanosecond() / TEN_POW[9 - self.fsp as usize];
-                    format!(
-                        "{}.{1:02$}",
-                        ZERO_DATETIME_NUMERIC_STR, nanos, self.fsp as usize
-                    )
-                } else {
-                    String::from(ZERO_DATETIME_NUMERIC_STR)
+    /// We assume that the `input` is trimmed and is not empty.
+    fn split_components(input: &str) -> Option<Vec<&[u8]>> {
+        let mut buffer = input.as_bytes();
+
+        debug_assert!(
+            !buffer.is_empty()
+                && !buffer.first().unwrap().is_ascii_whitespace()
+                && !buffer.last().unwrap().is_ascii_whitespace()
+        );
+
+        let mut components = Vec::with_capacity(7);
```

Comment:
Would you mind define constexpr like `7` using const variable and naming it?

Reply:
OK. THX~